### PR TITLE
Allow hashes and drops to interop more cleanly

### LINKED ISSUE #4326: Jekyll master _still_ not playing nicely with jekyll-archives
The latest bug in a continuing saga… (see https://github.com/jekyll/jekyll-archives/issues/55, and https://github.com/jekyll/jekyll/issues/4298)

I’m using the pre-release of Jekyll 3.1 (on master) with the latest liquid drop related fixes, but still getting an error, albeit different from the one before. Going by the trace, it still seems to relate to jekyll-archives. Here’s that trace:

```
Liquid Exception: no implicit conversion of nil into String in _includes/title.html, included in _layouts/index.html
/Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/tags/include.rb:132:in `rescue in render': no implicit conversion of nil into String (Jekyll::Tags::IncludeTagError)
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/tags/include.rb:124:in `render'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/block.rb:151:in `render_token'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/profiler/hooks.rb:5:in `block in render_token_with_profiling'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/profiler.rb:80:in `profile_token_render'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/profiler/hooks.rb:4:in `render_token_with_profiling'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/block.rb:135:in `block in render_all'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/block.rb:122:in `each'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/block.rb:122:in `render_all'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/block.rb:108:in `render'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/template.rb:210:in `block in render'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/template.rb:262:in `with_profiling'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/template.rb:209:in `render'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/template.rb:222:in `render!'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/liquid_renderer/file.rb:28:in `block (2 levels) in render!'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/liquid_renderer/file.rb:36:in `measure_bytes'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/liquid_renderer/file.rb:27:in `block in render!'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/liquid_renderer/file.rb:43:in `measure_time'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/liquid_renderer/file.rb:26:in `render!'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/convertible.rb:112:in `render_liquid'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/convertible.rb:216:in `render_all_layouts'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/convertible.rb:264:in `do_layout'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/jekyll-archives-2.1.0/lib/jekyll-archives/archive.rb:104:in `render'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/jekyll-archives-2.1.0/lib/jekyll-archives.rb:104:in `block in render'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/jekyll-archives-2.1.0/lib/jekyll-archives.rb:103:in `each'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/jekyll-archives-2.1.0/lib/jekyll-archives.rb:103:in `render'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/jekyll-archives-2.1.0/lib/jekyll-archives.rb:48:in `generate'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/site.rb:154:in `block in generate'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/site.rb:153:in `each'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/site.rb:153:in `generate'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/site.rb:58:in `process'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/command.rb:26:in `process_site'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/commands/build.rb:58:in `build'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/commands/build.rb:33:in `process'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/commands/build.rb:16:in `block (2 levels) in init_with_program'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/mercenary-0.3.5/lib/mercenary/command.rb:220:in `call'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/mercenary-0.3.5/lib/mercenary/command.rb:220:in `block in execute'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/mercenary-0.3.5/lib/mercenary/command.rb:220:in `each'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/mercenary-0.3.5/lib/mercenary/command.rb:220:in `execute'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/mercenary-0.3.5/lib/mercenary/program.rb:42:in `go'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/mercenary-0.3.5/lib/mercenary.rb:19:in `program'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/bin/jekyll:13:in `<top (required)>'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bin/jekyll:23:in `load'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bin/jekyll:23:in `<main>'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bin/ruby_executable_hooks:15:in `eval'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bin/ruby_executable_hooks:15:in `<main>'
```

The source files referenced can be seen here: https://github.com/paulrobertlloyd/paulrobertlloyd.com/blob/master/source/_includes/title.html and https://github.com/paulrobertlloyd/paulrobertlloyd.com/blob/master/source/_layouts/index.html

If I remove `{% include title.html %}` from /_layouts/index.html, I get a different error, still with reference to jekyll-archives:

```
  Liquid Exception: undefined method `[]' for nil:NilClass in _layouts/default.html
/Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/jekyll-assets-2.1.2/lib/jekyll/assets/liquid/tag.rb:80:in `render': undefined method `[]' for nil:NilClass (NoMethodError)
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/block.rb:151:in `render_token'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/profiler/hooks.rb:5:in `block in render_token_with_profiling'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/profiler.rb:80:in `profile_token_render'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/profiler/hooks.rb:4:in `render_token_with_profiling'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/block.rb:135:in `block in render_all'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/block.rb:122:in `each'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/block.rb:122:in `render_all'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/block.rb:108:in `render'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/template.rb:210:in `block in render'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/template.rb:262:in `with_profiling'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/template.rb:209:in `render'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/liquid-3.0.6/lib/liquid/template.rb:222:in `render!'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/liquid_renderer/file.rb:28:in `block (2 levels) in render!'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/liquid_renderer/file.rb:36:in `measure_bytes'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/liquid_renderer/file.rb:27:in `block in render!'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/liquid_renderer/file.rb:43:in `measure_time'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/liquid_renderer/file.rb:26:in `render!'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/convertible.rb:112:in `render_liquid'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/convertible.rb:216:in `render_all_layouts'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/convertible.rb:264:in `do_layout'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/jekyll-archives-2.1.0/lib/jekyll-archives/archive.rb:104:in `render'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/jekyll-archives-2.1.0/lib/jekyll-archives.rb:104:in `block in render'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/jekyll-archives-2.1.0/lib/jekyll-archives.rb:103:in `each'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/jekyll-archives-2.1.0/lib/jekyll-archives.rb:103:in `render'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/jekyll-archives-2.1.0/lib/jekyll-archives.rb:48:in `generate'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/site.rb:154:in `block in generate'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/site.rb:153:in `each'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/site.rb:153:in `generate'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/site.rb:58:in `process'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/command.rb:26:in `process_site'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/commands/build.rb:58:in `build'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/commands/build.rb:33:in `process'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/lib/jekyll/commands/build.rb:16:in `block (2 levels) in init_with_program'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/mercenary-0.3.5/lib/mercenary/command.rb:220:in `call'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/mercenary-0.3.5/lib/mercenary/command.rb:220:in `block in execute'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/mercenary-0.3.5/lib/mercenary/command.rb:220:in `each'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/mercenary-0.3.5/lib/mercenary/command.rb:220:in `execute'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/mercenary-0.3.5/lib/mercenary/program.rb:42:in `go'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/gems/mercenary-0.3.5/lib/mercenary.rb:19:in `program'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bundler/gems/jekyll-50086b686a89/bin/jekyll:13:in `<top (required)>'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bin/jekyll:23:in `load'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bin/jekyll:23:in `<main>'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bin/ruby_executable_hooks:15:in `eval'
	from /Users/paulrobertlloyd/.rvm/gems/ruby-2.2.1/bin/ruby_executable_hooks:15:in `<main>'
```

Any ideas?

### PR DESCRIPTION
**Pull request by pathawks**
~~This fixes #4326 and doesn't break anything else.~~

There has *got* to be a better way...

### CODE REVIEW

File: lib/jekyll/utils.rb [hunk 1]
```diff
@@ -34,6 +34,7 @@ def deep_merge_hashes(master_hash, other_hash)
     # Thanks to whoever made it.
     def deep_merge_hashes!(target, overwrite)
       overwrite.each_key do |key|
+        next if overwrite[key].nil? && target.key?(key)
```
Comment (parkr) [inline lib/jekyll/utils.rb]: This makes it impossible to nil-ify a key :(

Comment (envygeeks) [inline lib/jekyll/utils.rb]: A tough problem indeed, however, why would you want a nil key?

Comment (parkr) [inline lib/jekyll/utils.rb]: If the target has a value and you want to get rid of it.

**Comment by pathawks** [comment]
This test was checking the original hash instead of the merged hash. That's why the test was not failing as it should have.

File: lib/jekyll/utils.rb (hunk 1 above)
Comment (pathawks) [inline lib/jekyll/utils.rb]: > This makes it impossible to nil-ify a key :(

The problem is that the drop will *always* have a `page` key, so what is the other option?

**Comment by envygeeks** [comment]
Hey @pathawks, this is a heads up that your pull has caused a failure on JRuby and while right now it's not important soon I will be asking @jekyll/core to make JRuby first-class, mind checking it out? If you need help I'm happy to help.

**Comment by envygeeks** [comment]
Nevermind, it seems the error rotates every time I re-run the tests o.O don't worry about it.

**Comment by parkr** [comment]
I like Jordon's fix here: https://github.com/jekyll/jekyll/issues/4326#issuecomment-170320774

**Comment by pathawks** [comment]
Should I revert that commit and just leave the failing test?

**Comment by parkr** [comment]
> Should I revert that commit and just leave the failing test?

Did you try using just Jordon's refactor?

**Comment by pathawks** [comment]
> Did you try using just Jordon's refactor?

Yes, it certainly does not work as is, and I'm not smart enough to figure out what the problem is.

**Comment by envygeeks** [comment]
@parkr @pathawks you need to adjust Jekyll Drops to add more hash methods via Forwardable.

**Comment by envygeeks** [comment]
The premise of that speed up is optimizing both Jekyll Drops and Utils so that Utils doesn't need repath for Drops and Drops provides access to it's enumerators for Jekyll.

File: lib/jekyll/utils.rb [hunk 2]
```diff
@@ -28,26 +28,16 @@ def deep_merge_hashes(master_hash, other_hash)
...
-        end
-
-        target[key] = overwrite[key]
-      end
-
-      if target.respond_to?(:default_proc) && overwrite.respond_to?(:default_proc) && target.default_proc.nil?
-        target.default_proc = overwrite.default_proc
+    # Returns the merged hashes.
+    def deep_merge_hashes!(old_hash, new_hash)
+      old_hash.merge(new_hash) do |_, old_val, new_val|
```
Comment (parkr) [inline lib/jekyll/utils.rb]: This should be `merge!` I think, as this is `deep_merge_hashes!`.

**Pull request closed without merging**